r/R/dplyr-summarize.R

```diff
@@ -322,15 +301,76 @@ arrow_eval_or_stop <- function(expr, mask) {
   out
 }
 
+# This function returns a list of expressions that can be used to project the
+# data before an aggregation to only the fields required for the aggregation,
+# including the fields used in the aggregations (the "targets") and the group
+# fields. The names of the returned list are used to ensure that the projection
+# node is wired up correctly to the aggregation node.
 summarize_projection <- function(.data) {
   c(
-    map(.data$aggregations, ~ .$data),
+    unlist(unname(imap(
+      .data$aggregations,
+      ~set_names(
+        .x$data,
+        aggregate_target_names(.x$data, .y)
+      )
+    ))),
+    .data$selected_columns[.data$group_by_vars]
+  )
+}
+
+# This function determines what names to give to the fields used in aggregations
+# (the "targets"). When an aggregate function takes 2 or more fields as targets,
+# this function gives the fields unique names by appending `..1`, `..2`, etc.
+# When an aggregate function is nullary, this function returns a zero-length
+# character vector.
+aggregate_target_names <- function(data, name) {
+  if (length(data) > 1) {
+    paste(name, seq_along(data), sep = "..")
+  } else if (length(data) > 0) {
+    name
+  } else {
+    character(0)
+  }
+}
+
+# This function returns a list of expressions representing the aggregated fields
+# that will be returned by an aggregation
+aggregated_fields <- function(aggs) {
+  map(
+    aggs,
+    ~Expression$create(.$fun, args = .$data, options = .$options)
+  )
+}
+
+# Unlike with other pairs of non-hash/hash aggregate kernels in the Arrow C++
+# library, the `tdigest` and `hash_tdigest` kernels have different output types.
+# The `tdigest` kernel returns `Float64`, but the `hash_tdigest` kernel returns
+# `FixedSizeList[Float64]`. The system that the R bindings use to infer the
+# output types of expressions does not account for this. It infers the output
+# type of `hash_tdigest` as `Float64`. This function is used to correct this.
+fix_aggregated_types <- function(fields, aggs, hash) {
+  imap(
+    fields,
+    ~if (hash && aggs[[.y]]$fun == "tdigest") {
+        fixed_size_list_of(float64(), 1L)
```

Comment:
Is it always float64()? (looks like it is today, but probably not safe to assume that things like timestamp wouldn't return timestamp)

Is it always length 1? (looks like not, the length depends on the number of quantiles you give it in the options)

I think this would generalize better:

```suggestion
        fixed_size_list_of(.x, length(aggs[[.y]][["options"]][["q"]])
```

Reply:
Nevermind, see above, I think we can construct `hash_` Expressions and get the actual output type without any munging.